src/algebra/order/hom/monoid.lean

```diff
@@ -0,0 +1,481 @@
+instance order_monoid_hom_class.to_order_hom_class [order_monoid_hom_class F α β] :
+  order_hom_class F α β :=
+{ map_rel := order_monoid_hom_class.monotone }
+
+@[to_additive]
+instance [order_monoid_hom_class F α β] : has_coe_t F (α →*o β) :=
+⟨λ f, { to_fun := f, map_one' := map_one f, map_mul' := map_mul f,
+  monotone' := order_monoid_hom_class.monotone _ }⟩
+
+end monoid
+
+section monoid_with_zero
+variables [linear_ordered_comm_monoid_with_zero α] [linear_ordered_comm_monoid_with_zero β]
+
+/-- `order_monoid_with_zero_hom α β` is the type of functions `α → β` that preserve
+the `monoid_with_zero` structure.
+
+`order_monoid_with_zero_hom` is also used for group homomorphisms.
+
+When possible, instead of parametrizing results over `(f : α →+ β)`,
+you should parametrize over `(F : Type*) [order_monoid_with_zero_hom_class F α β] (f : F)`.
+
+When you extend this structure, make sure to extend `order_monoid_with_zero_hom_class`. -/
+structure order_monoid_with_zero_hom (α β : Type*)
+  [linear_ordered_comm_monoid_with_zero α] [linear_ordered_comm_monoid_with_zero β]
+  extends α →*₀ β :=
+(monotone' : monotone to_fun)
+
+infixr ` →*₀o `:25 := order_monoid_with_zero_hom
+
+/-- `order_monoid_with_zero_hom_class F α β` states that `F` is a type of
+ordered monoid with zero homomorphisms.
+
+You should also extend this typeclass when you extend `order_monoid_with_zero_hom`. -/
+class order_monoid_with_zero_hom_class (F : Type*) (α β : out_param $ Type*)
+  [linear_ordered_comm_monoid_with_zero α] [linear_ordered_comm_monoid_with_zero β]
+  extends monoid_with_zero_hom_class F α β :=
+(monotone (f : F) : monotone f)
+
+@[priority 100] -- See note [lower instance priority]
+instance order_monoid_with_zero_hom_class.to_order_monoid_hom_class
+  [order_monoid_with_zero_hom_class F α β] : order_monoid_hom_class F α β :=
+{ .. ‹order_monoid_with_zero_hom_class F α β› }
+
+instance [order_monoid_with_zero_hom_class F α β] : has_coe_t F (α →*₀o β) :=
+⟨λ f, { to_fun := f, map_one' := map_one f, map_zero' := map_zero f, map_mul' := map_mul f,
+  monotone' := order_monoid_with_zero_hom_class.monotone _ }⟩
+
+end monoid_with_zero
+
+namespace order_monoid_hom
+section
+variables [ordered_comm_monoid α] [ordered_comm_monoid β] [ordered_comm_monoid γ]
+  [ordered_comm_monoid δ] {f g : α →*o β}
+
+@[to_additive]
+instance : order_monoid_hom_class (α →*o β) α β :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by { obtain ⟨⟨_, _⟩, _⟩ := f, obtain ⟨⟨_, _⟩, _⟩ := g, congr' },
+  map_mul := λ f, f.map_mul',
+  map_one := λ f, f.map_one',
+  monotone := λ f, f.monotone' }
+
+
+@[ext, to_additive] lemma ext (h : ∀ a, f a = g a) : f = g := fun_like.ext f g h
+@[to_additive] lemma ext_iff : f = g ↔ ∀ a, f a = g a := fun_like.ext_iff
+@[to_additive] lemma congr_fun (h : f = g) (a : α) : f a = g a := fun_like.congr_fun h a
+@[to_additive] lemma coe_inj ⦃f g : α →*o β⦄ (h : (f : α → β) = g) : f = g :=
+fun_like.coe_injective h
```

Comment:
Only an `ext` lemma is really required here (due to limitations in the `ext` tactic), for the rest we don't need to duplicate what `fun_like` already provides.
```suggestion
@[ext, to_additive] lemma ext (h : ∀ a, f a = g a) : f = g := fun_like.ext f g h
```

Reply:
Good point, I'll go through and deprecate these (at least, write a docstring pointing out you should use the `fun_like` stuff instead.)